relay-server/src/services/project.rs

```diff
@@ -932,9 +932,24 @@ impl Project {
             Ok(current_limits.check_with_quotas(quotas, item_scoping))
         });
 
-        let (enforcement, mut rate_limits) =
-            envelope_limiter.enforce(envelope.envelope_mut(), &scoping)?;
-        enforcement.track_outcomes(envelope.envelope(), &scoping, outcome_aggregator);
+        let (mut enforcement, mut rate_limits) =
+            envelope_limiter.compute(envelope.envelope_mut(), &scoping)?;
+
+        let check_nested_spans = state
+            .as_ref()
+            .is_some_and(|s| s.has_feature(Feature::ExtractSpansFromEvent));
+
+        // If we can extract spans from the event, we want to try and count the number of nested
+        // spans to correctly emit negative outcomes in case the transaction itself is dropped.
+        if check_nested_spans {
+            let spans_count = count_nested_spans(&envelope);
+            if spans_count > 0 {
+                add_nested_span_enforcements(&mut enforcement, spans_count);
+            }
```

Comment:
I'd call `count_nested_spans` from within `add_nested_span_enforcements` (which may need a new name then).

The high level control flow is `if feature_enabled { sync_spans_to_enforcement() }`, how to count the spans ins the concern of the sync function.